### PR TITLE
Explore test translation

### PR DESCRIPTION
**Pull request by marler8997**
An initial draft to implement @atilaneves 's idea to support autowrap tests for multiple languages.  The idea is to write the tests once in python and translate them to other languages.  Since python can be easily parsed, it's a good candidate to write the test source in.

Note that if there are language-specific tests that can't be represented in python, we'll likely want to have a mix of common tests written in python and language-specific tests.

### CODE REVIEW

**Comment by marler8997** [comment]
Here are @atilaneves initial comments:


> run_translate_tests shouldn't depend on being run from the directory where it is. In general, no bash script should. The first thing I did was run tests/run_translate_tests and that broke.

Agreed.  I was just using this script to run the tool personally, I was either going to remove it before making a PR or enhance it depending on if we wanted to test the translation tool or not.

> I'm not sure why example.py is there? It doesn't have any tests in it.

I've renamed it to `exercise_ast_to_ir.py`.  The point of if was to generate corner cases in the AST to make sure my IR data structures were designed correctly to handle all the AST variants.

> The generated C# code should be runnable as a test. Copying the test_simple_cs.csproj should probably do, as would adding a Makefile rule

Ok I've modified it a bit to generate the proper C# code to make it a valid Microsoft Visual Studio unit test.

> Please pep8ify all the Python code

I ran pycodestyle on the new pthon files.  I fixed the errors except for one that I"m not sure how to fix (see code).  Is this the tool I should be using to verify it is pep8ifieid?

> I have comments on the Python code but I'd rather write them on a PR where it's linked to line numbers

I've created this Draft PR so you can comment.

**Comment by atilaneves** [comment]
> I ran pycodestyle on the new python files. I fixed the errors except for one that I"m not sure how to fix (see code). Is this the tool you'd like me to use to verify it is pep8ifieid?

I'm not familiar with pycodestyle. I meant [this pip package](https://pypi.org/project/pep8ify/). I don't even know if that's the recommended way these days. Since PEP8 violations trigger pyflakes as I type in the editor, I basically never check in any files not in PEP8 anyway, so I haven't had to use a tool to enforce it.

File: tests/pytest_translate [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+#
+# Translate a python test to another language.
+#
+import enum
+import sys
+import os
+import argparse
+import ast
+
+sys.path.insert(0, os.path.join(os.path.dirname(
```
Comment (atilaneves) [inline tests/pytest_translate]: Modifying `sys.path` from code is nearly always a code smell.

File: tests/pytest_translate [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    os.path.abspath(__file__)), "translate"))
+
+# Not sure how to solve thie pep8 error:
+#   pytest_translate:17:1: E402 module level import not at top of file
+# this cannot be imported until we add the import path above
+import pythonir
+
+parser = argparse.ArgumentParser(
```
Comment (atilaneves) [inline tests/pytest_translate]: It'd be nice if there were hardly any file-scope statements. I'm aware this is common for scripts, but the likelihood that this code will grow beyond a simple script is incredibly high. It'd be nice if the script part of the code was just calling a simple main function.

File: tests/pytest_translate [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+parser.add_argument("output_file")
+
+args = parser.parse_args()
+
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+
+eprint(f"input_file: {args.input_file}")
```
Comment (atilaneves) [inline tests/pytest_translate]: Do we really need this?

File: tests/pytest_translate [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    PYTHON = 0
+    CSHARP = 1
+
+
+def get_fileext_language(filename):
+    if filename.endswith(".py"):
+        return Language.PYTHON
+    if filename.endswith(".cs"):
+        return Language.CSHARP
+    return None
```
Comment (atilaneves) [inline tests/pytest_translate]: This should probably raise an exception instead.

File: tests/pytest_translate [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+
+output_language = get_fileext_language(args.output_file)
+if not output_language:
+    sys.exit(f"ERROR: unknown file extension from '{args.output_file}'")
+eprint(f"output_language: {output_language}")
+
+
+class IrGenVisitor(ast.NodeVisitor):
```
Comment (atilaneves) [inline tests/pytest_translate]: I think the visitor class shouldn't be in the same file as the script code but in an imported module.

File: tests/pytest_translate [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __init__(self):
+        self.ast_depth = 0
+        self.current_scope = None
+        self.scope_depth = 0
+
+    def log(self, msg):
+        print(f"{self.ast_depth*' '}{msg}")
+
+    def visitimpl(self, node):
```
Comment (atilaneves) [inline tests/pytest_translate]: `_visitimpl`?

File: tests/pytest_translate [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        parent_scope = self.current_scope
+        # args, body, name, returns
+        func = pythonir.Function(node.name)
+        self.visit_new_scope(node, func)
+        self.current_scope.add_function_def(func)
+
+    def visit_Assign(self, node):
+        assert(self.current_scope)
+        self.log(f"visitAssign targets={node.targets} value={node.value}")
+        # self.current_scope.add_assign(node)
```
Comment (atilaneves) [inline tests/pytest_translate]: ?

File: tests/pytest_translate [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        self.visitimpl(node)
+
+
+module = IrGenVisitor().visit(input_ast)
+
+if output_language == Language.CSHARP:
+    import csharpgen
+    csharpgen.translate(module, args.output_file)
+else:
+    sys.exit(f"unsupported output language '{output_language}'")
```
Comment (atilaneves) [inline tests/pytest_translate]:
```python
# here
import codegen
codegen.translate(module, output_langage, output_file)
```

```python
# codegen.py
def translate(module, output_language, output_file):
    if output_language != Language.CSHARP:
        raise NotImplementedError(f"Unsupported output language '{output_language}'")
    else:
        return translate_cs(module, output_file)
```

File: tests/run_translate_tests [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (atilaneves) [inline tests/run_translate_tests]: The main issue I have with this file is that it doesn't actually check anything output the output, nor is integrated with the current Makefile at the root of the project.

File: tests/translate/csharpgen.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        self.writeln("}")
+
+
+def translate(module, filename):
+    with open(filename, "w") as file:
+        gen = CSharpGenerator(file)
+        translate_to(module, gen)
+
+
+def translate_to(module, gen):
```
Comment (atilaneves) [inline tests/translate/csharpgen.py]: Why is this not a `CharpGenerator` method?

File: tests/translate/csharpgen.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    # first get all imports
+    # then generate project/assembly references and/or comments in the
+    # generated code
+    # for now just hard-code them
+    # Note that these will change depending on the .NET version we are targeting
+    references = [
```
Comment (atilaneves) [inline tests/translate/csharpgen.py]: references?

File: tests/translate/csharpgen.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    top_level_exec_nodes = []
+
+    # we use the fully-qualified names to avoid name-collisions with the symbols from the test
+    gen.writeln(f"[Microsoft.VisualStudio.TestTools.UnitTesting.TestClass]")
+    gen.writeln("public class TestMain")
+    gen.open_block()
+    # TODO: should I use a visitor for this?
+    for node in module.nodes:
+        if isinstance(node, pythonir.Function):
+            # TODO: this code will be moved into the Function visitor and will
```
Comment (atilaneves) [inline tests/translate/csharpgen.py]: Good, `isinstance` is nearly a code smell.

File: tests/translate/csharpgen.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            #       work for both free functions and member functions
+            #
+            # TODO: how do we tell test methods apart from normal methods?
+            #       if they start with "test_"?
+            gen.writeln(f"[Microsoft.VisualStudio.TestTools.UnitTesting.TestMethod]")
+            gen.writeln(f"public void {node.name}()")
+            gen.open_block()
+            # node.generate_csharp(gen)
```
Comment (atilaneves) [inline tests/translate/csharpgen.py]: ?

File: tests/translate/csharpgen.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    #       an error
+    if len(top_level_exec_nodes) > 0:
+        print("ERROR: the python test has top-level exec nodes?")
+    # gen.writeln("public static void Main(System.String[] args)")
+    # gen.open_block()
+    # gen.writeln(
+    #     "Console.WriteLine(\"TODO: finish generating test code...\");")
+    # for node in top_level_exec_nodes:
+    #     node.generate_csharp(gen)
+    # gen.close_block()
```
Comment (atilaneves) [inline tests/translate/csharpgen.py]: ?

File: tests/translate/exercise_ast_to_ir.py [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# Only meant to test the code that converts python AST to python IR
```
Comment (atilaneves) [inline tests/translate/exercise_ast_to_ir.py]: Shouldn't this be part of a unit test instead then?

File: tests/translate/pythonir.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+            return self.name
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class Import:
+    def __init__(self, module, symbols):
+        self.module = module
+        self.symbols = symbols
```
Comment (atilaneves) [inline tests/translate/pythonir.py]: These two classes don't have a base class or child classes so I'm not sure why they exist. If they are data classes I suggest using that instead.

File: tests/translate/pythonir.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
+# TODO: I think I'll need this BaseNode class but not sure yet
```
Comment (atilaneves) [inline tests/translate/pythonir.py]: YAGNI.

Comment (atilaneves) [reply tests/translate/pythonir.py]: `pythonir` seems like an odd module name to me. `python` as opposed to...?

**Comment by atilaneves** [comment]
> I've renamed it to exercise_ast_to_ir.py. The point of if was to generate corner cases in the AST to make sure my IR data structures were designed correctly to handle all the AST variants

The problem is that the exercise Python file isn't a pytest test file. It doesn't make sense to try and translate it to C# (or to anything else). To test the IR parts of the code, please write unit tests runnable from pytest.

File: tests/pytest_translate (hunk 1 above)
Comment (marler8997) [reply tests/pytest_translate]: What's the alternative here?  How do I import modules in another path?

File: tests/pytest_translate (hunk 7 above)
Comment (marler8997) [reply tests/pytest_translate]: something that will need to be implemented later, I commented it out for now so I didn't develop too much code without review

File: tests/pytest_translate (hunk 8 above)
Comment (marler8997) [reply tests/pytest_translate]: I think this will depend on how much each language shares in code.  For now each language has it's own codegen module so they don't even need the `Language` enum.

File: tests/translate/pythonir.py (hunk 17 above)
Comment (marler8997) [reply tests/translate/pythonir.py]: it's meant to hold IR that represents the python test.  But I've renamed it to `ir.py`.

File: tests/translate/pythonir.py (hunk 16 above)
Comment (marler8997) [reply tests/translate/pythonir.py]: Oh I've never heard of a dataclass, thanks for the tip.

File: tests/pytest_translate (hunk 5 above)
Comment (marler8997) [reply tests/pytest_translate]: moved to `ir.py`

File: tests/pytest_translate (hunk 1 above)
Comment (atilaneves) [reply tests/pytest_translate]: The usual Python way, by creating a package with `__init__.py`. I wouldn't put the implementation in the tests directory anyway, I'd create a new top-level directory in the repo, put all of the translation files there, and just do regular imports. 

Now, tests to the translation code could and should be located in the tests directory. And obviously the output of such translations.

File: tests/pytest_translate (hunk 3 above)
Comment (marler8997) [reply tests/pytest_translate]: nope, removed

File: tests/translate/csharpgen.py (hunk 11 above)
Comment (marler8997) [reply tests/translate/csharpgen.py]: I've renamed it to `assembly_references`.  These will need to be infered from the imports but for now I'm just hard-coding them.

File: tests/run_translate_tests (hunk 9 above)
Comment (marler8997) [reply tests/run_translate_tests]: I've integrated it with the Makefile.  This file is just running the translation program and checking the exit code (via the `set -e`).  So just making sure the program successfully runs.

File: translate/ir.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
+from dataclasses import dataclass
+import typing
+import ast
+
+
+@dataclass
+class ImportedSymbol:
+    name: str
```
Comment (marler8997) [inline translate/ir.py]: Not sure what to do here.  `pep8ify` translates this to:
```python
class ImportedSymbol:
    name:
        str
    alias:
        str
```

But then I get a syntax error if I try to run it:
```
Traceback (most recent call last):
  File "/home/marler8997/git/autowrap/translate/test_ir.py", line 3, in <module>
    import ir
  File "/home/marler8997/git/autowrap/translate/ir.py", line 8
    name:
        ^
SyntaxError: invalid syntax
```

So I'm having to manually fix it to be on one line after I run `pep8ify`.

Comment (atilaneves) [reply translate/ir.py]: Argh, tools.

File: translate/asttoirvisitor.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        self.ast_depth += 1
+        ast.NodeVisitor.generic_visit(self, node)
+        self.ast_depth -= 1
+
+    def _visit_new_scope(self, node, new_scope):
+        parent_scope = self.current_scope
+        self.current_scope = new_scope
+        self.scope_depth += 1
+        self._visitimpl(node)
+        assert(new_scope == self.current_scope)
```
Comment (atilaneves) [inline translate/asttoirvisitor.py]: The parentheses aren't needed.

**Comment by atilaneves** [comment]
Minor nitpick aside (assert with parens), should be good to go.

**Comment by marler8997** [comment]
Fixed, parens removed from assert statements.

File: Makefile [hunk 20]
```diff
@@ -22,7 +22,7 @@ export PYD_D_VERSION_13 ?= Python_3_8_Or_Later
 .PHONY: clean test test_python test_python_pyd test_python_pynih test_cs test_simple_pyd test_simple_pynih test_simple_cs test_pyd_pyd test_issues_pyd test_issues_pynih test_numpy_pyd test_numpy_pynih examples/simple/lib/pyd/libsimple.so examples/simple/lib/pynih/libsimple.so examples/issues/lib/pyd/libissues.so examples/issues/lib/pynih/libissues.so examples/numpy/lib/pyd/libnumpy.so examples/numpy/lib/pynih/libnumpy.so examples/pyd/lib/pyd/libpydtests.so examples/pyd/lib/pynih/libpydtests.so examples/phobos/lib/pyd/libphobos.so examples/phobos/lib/pynih/libphobos.so examples/simple/lib/csharp/libsimple.so examples/simple/Simple.cs
 
 all: test
-test: test_python test_cs
+test: translate_tests test_python test_cs
```
Comment (atilaneves) [inline Makefile]: translate_tests has to be a .PHONY target

**Comment by atilaneves** [comment]
I'm going to merge as-is then create a PR myself and add you as a reviewer.
